kitchen/src/kitchen/demos.clj

```diff
@@ -0,0 +1,144 @@
+      (utils/exception->json-response e))))
+
+(defn- directory-location
+  [dir-name]
+  (let [base-dir (or (System/getenv "MESOS_DIRECTORY") (System/getProperty "user.dir"))]
+    (str base-dir File/separator dir-name)))
+
+(defn- copy-uri-to-file
+  [uri file]
+  (with-open [in (io/input-stream uri)
+              out (io/output-stream file)]
+    (io/copy in out)))
+
+(let [image-tagging-lock (Object.)]
+  (defn- retrieve-image-tags
+    [classifier-file image-location predictions]
+    (let [result-atom (atom nil)]
+      (locking image-tagging-lock
+        (reset! result-atom (sh/sh "python3"
+                                   classifier-file
+                                   (str "--image_file=" image-location)
+                                   (str "--num_top_predictions=" predictions))))
+      @result-atom)))
+
+(defn- image-tagging-handler
+  "Performs image tagging using tensorflow."
+  [{:keys [query-params] :as request}]
+  (printlog request "query params:" query-params)
+  (try
+    (let [{:strs [id name url predictions]} query-params
+          image-location (directory-location (str "images" File/separator name))
+          classifier-file-name "classify_image.py"
+          classifier-file (directory-location (str "images" File/separator classifier-file-name))]
+
+      (-> image-location str File. .getParentFile .mkdirs)
+      (-> classifier-file str File. .getParentFile .mkdirs)
+
+      (printlog request "downloading" url "to" image-location)
+      (copy-uri-to-file url image-location)
+
+      (printlog request "classifier resource:" (io/resource classifier-file-name))
+      (when-not (-> classifier-file str File. .exists)
+        (printlog request "saving" classifier-file-name "to" classifier-file)
+        (copy-uri-to-file (io/resource classifier-file-name) classifier-file))
+
+      (printlog request "classifying image" image-location)
+      (let [{:keys [err exit out]} (retrieve-image-tags classifier-file image-location predictions)
+            result-lines (filter #(str/includes? % "(score =") (str/split-lines out))]
+        (printlog request "classifier exit code:" exit)
+        (printlog request "classifier stdout:" out)
+        (printlog request "classifier stderr:" err)
+        (utils/map->json-response (cond-> {:id id, :out out, :predictions result-lines}
+                                          (not (str/blank? err))
+                                          (assoc :err err))
+                                  :status (if (zero? exit) 200 500))))
+    (catch Exception e
+      (log/error e "encountered exception while tagging image")
+      (utils/exception->json-response e))))
+
+(defn demo-handler
+  [{:keys [uri] :as request}]
+  (case uri
+    "/demos/image-search" (image-search-handler request)
+    "/demos/image-tagging" (image-tagging-handler request)
+    {:status 200
+     :headers {"content-type" "text/html"}
+     :body (io/input-stream (io/resource "demo.html"))}))
```

Comment:
Does this just happen when you hit /demos? If so, shouldn't we make the route specific to this demo?

Reply:
I left it as the default as we have a single html file until we add another demo